### DISCUSSION TITLE
How does consent for individual vendors work?

### DISCUSSION THREAD

**Discussion by bocops** [bocops/UMP-workarounds#10] [Q&A]
This is a continuation of the discussion started in #4, with the last question:

> About the instruction here of setting goodVendorConfiguration based on what I get from tcVendorString:
https://github.com/bocops/UMP-workarounds/issues/4#issuecomment-1860023070
> 
> I copied the long string there. It has multiple "1" there for some reason. Not just a single one. That's even though on this app I use only Admob.
> 
> If I close/dismiss the dialog, or choose "do not consent", it's just "0", and yet the function returns "unclear".
Is it really unclear? Shouldn't it be limited or non-personalized?
We can at least say that it's not "personalized" for sure, no?
Maybe you should add a new function, of detection of "do no consent" ?
> 
> I also don't know what happens if the user configures specific vendors there. There are many options for the users.
Maybe possible to have different ad-types for different vendors...
I can't find how to see the ID of all vendors, and in your code there is no mention of 755. Only this string that I need to set.

The starting point of all of this actually is your own AdMob settings. If you go to the "GDPR settings" in "Privacy & messaging":

![image](https://github.com/bocops/UMP-workarounds/assets/29770778/e0b8b2ba-2529-4a5a-91fb-78dc1eb68628)

you will find a section where you can "Review your ad partners":

![image](https://github.com/bocops/UMP-workarounds/assets/29770778/69b2cedd-fcba-4b3e-a65f-f43efc6acd14)

Note that in this case, there's only "1 partner", because I customized those settings for myself. It means that my app will only be able to display ads by that single ad partner. If instead I used "Commonly used ad partners", that would be 191 of them. Even adding just the "mediation partners who have been GDPR certified with Google" means that there are 37 of them.

Whatever you define there will make up the list of "vendors" that will be displayed in the UMP dialog of your app.

Now, in the UMP dialog, you can either:
1. On the first screen, click "Consent"
2. On the first screen, click "Manage options"; then on the second screen screen, click "Accept all"
3. On the first screen, click "Manage options"; then on the second pick individual choices and click "Confirm choices"
4. On the first screen, click "Manage options"; then on the second pick individual choices and scroll down to click "Vendor preferences"; then on that third screen pick individual vendor choices and click "Confirm choices", or just click "Accept all"

Cases 1 and 2 (and perhaps 4 with clicking "Accept all", unsure about that) are unproblematic, because a user will have given full consent for your app and everything is fine.

However, the problem is that when users just click "Confirm choices" on the secondary screen without selecting any options (case 3), or even if they selected any options but not any vendors on the tertiary screen (case 4, first part), **your app will not have sufficient consent to display ads** - likely none, but at least not from specific vendors that you _wanted_ to use, or else you wouldn't have selected them in AdMob in the first place.

Now, regarding the string that functionality in here wants you to copy&paste - that **is** the bit string for vendor consent you gave, so it needs to be copied after you've given full consent by clicking "Consent" or some variant of "Accept all". After full consent, every '1' character corresponds to a vendor that _you_ set up in AdMob before, and every '0' corresponds to a vendor that you did _not_ want to use. I have no say in that, all of it is based purely on your own choices.

Obviously, yes, if you close the UMP dialog and give no consent at all, the vendor bit string will be just a single '0', because it needs to contain no '1' characters in that case.

There is no explicit mention of a "755" in my code, because there's no need for it. That vendor ( which happens to be "Google Advertising Products") just becomes a part of the vendor bit string that you need to put into the function, with a '1' character in position 755 of the string.

For what it is worth, you can disable that check by just using `val goodVendorConfiguration = "0"` - but that obviously means that the code will no longer detect if consent for the necessary vendors is _not_ present, which means that it might return a positive value like `AdConfiguration.PERSONALIZED` if in fact the app will not be able to display any ads. This is _not_ a bug of my code, but the actual core idea of the whole function.

**Comment by AndroidDeveloperLB** [comment]
But if you know that "0" means that there was no consent given for anything, why return "unclear" ? 

As for "755", I said this because it was mentioned here, and also on StackOverflow, and there it has a function of "canShowPersonalizedAds" , which seems to be able to detect the difference here, yet as you said I will need to check it better because you said they only check about Admob which has the ID of 755.
I thought that if I use Admob alone it's the only one that provides ads, but after checking the string, it seems a lot exist there, so perhaps Admob works with various publishing organizations and each can work differently with the data... It's like mediation, but inside and without any configuration unless I want...

How can I change your code (or add a new function) to detect that user has chosen "Do not consent" (or close the dialog), then?

**Comment by bocops** [comment]
> But if you know that "0" means that there was no consent given for anything, why return "unclear" ?

Because I assume but do not know for sure, beyond any doubt, what happens in that scenario. You can just consider `UNCLEAR` to be the same as `NONE` if you like, but I think we've been through that already.

> How can I change your code (or add a new function) to detect that user has chosen "Do not consent" (or close the dialog), then?

<s>The code does not need to be changed for that, because it already returns `NONE` in case of insufficient consent for anything (which includes closing and "Do not consent") **before** it attempts the vendor consent check.</s>

**Comment by AndroidDeveloperLB** [comment]
But as I said: After showing the dialog (I removed the app before checking), and calling the function when reaching the dialog callback, I get "Unclear", even though I chose "do not consent"...
That's because  tcInterestString is "01000011111" (and tcConsentString is "0") in this case, so "sufficientInterest" check is skipped and so it doesn't return "None".

So it does reach the vendors part

**Reply by bocops** [reply]
OK, I just checked and that is true. I edited my above reply accordingly.

However, the other part of my reply still stands: It does reach the vendor check because "Do not consent" apparently no longer sets all "legitimate interest" to false - which, I believe, a previous version of the UMP dialog did. This means that we do have "enough" for the return value to otherwise be `LIMITED`, but the fact that there's no vendor information makes it `UNCLEAR` what really happens.

I'm thinking of a way to return all that information to the caller. This would probably have to be more than a single enum value, because there's just too much information. For what it is worth, I'm also considering making these individual files an actual library that others could use in their app, but there would need to be decent interest in that.

In any case, if you want to edit the function for your own needs, if you want to find out if the user clicked "Do not consent", that seems to be the case if both the consent string and the vendor string (but not necessarily the interest string) are exactly "0".

**Comment by AndroidDeveloperLB** [comment]
I changed your code only a little bit, by moving parts of the code to where they are used, meaning as such:

[file.zip](https://github.com/bocops/UMP-workarounds/files/13742933/file.zip)

**Comment by AndroidDeveloperLB** [comment]
So I guess that if the vendor variable is "0", it's "do not consent".
I wonder if there are more cases that should be fixed. 
If we use mediation, what would it mean for the vendors variable? Also need to update and change it each time we add/remove some ad-network to handle there?

Also, how come you don't check the pref of "IABTCF_VendorLegitimateInterests", as on the StackOverflow answer?
